Generate a website with docs via `mkdocs`

ilyagr opened the pull request:
This is more or less ready for review now.

You can play with the website this generates on https://ilyagr.github.io/jj (which redirects to a placeholder docs for a "released" version of jj) and https://ilyagr.github.io/jj/prerelease. 

This includes a github action that updates the "prerelease" section of the website on every update to main.

I initially was thinking of using `mdbook`, which looks a little better, but I think versioning the docs is important, and the features I want are pretty much only supported by the Mkdocs' "Material" theme. For example, I would have preferred to use the [readthedocs theme](https://www.mkdocs.org/user-guide/choosing-your-theme/#readthedocs) if it could keep the URL when switching versions (https://github.com/jimporter/mike/issues/81).

Mkdocs is written in Python. The prerequesites for building docs on your machine should be to install Python and Poetry, everything else should be installed automatically by Poetry. See contributing.md for instructions.

(Mostly) fixes #1430.

PhilipMetzger commented on docs/github.md:
```diff
@@ -76,6 +76,12 @@ commits to your branch[^1]. Some projects (such as Jujutsu and LLVM) instead
 prefer that you keep your commits clean by rewriting them and then
 force-pushing[^2].
 
+[^1]: This is a GitHub Style review, as GitHub currently only is able to compare
+branches.
+
+[^2]: If you're wondering why we prefer clean commits in this project, see
+e.g. [this blog post][stacked]
+
```
Optional: Move all links to EOF?

ilyagr replied:
Thanks! This made a difference with `mdbook`, but there's no point to this change anymore.

PhilipMetzger commented on docs/install-and-setup.md:
SInce this is a new file, directly move to `shell` code blocks.

ilyagr replied:
I'd rather not. It's unlikely to matter, but that could make git's job easier in detecting the move and/or make fixing any merge conflicts easier later.

PhilipMetzger commented on docs/contributing.md:
```diff
@@ -131,7 +134,57 @@ These are listed roughly in order of decreasing importance.
+Once you have `poetry` installed, you should ask it to install the rest
+of the required tools into a virtual environment as follows:
+
+    poetry install
```
For consistency with other files, use a single backquoted block here instead of 4 spaces.

PhilipMetzger commented on docs/contributing.md:
```diff
@@ -184,6 +189,65 @@ occasionally check the terminal from which you ran `mkdocs serve` for
+[`mike`](https://github.com/jimporter/mike), which can be run with
+`poetry run -- mike`.
+
+On a POSIX system, one way to build the entire website is as follows
```
Minor suggestion: On a POSIX system or WSL, one way to build the entire website is as follows.

ilyagr replied:
Thanks!

PhilipMetzger reviewed (approved):
This is great. I've left some nits.

ilyagr commented:
Thank you Philip! I addressed your comments, added some more details to `contributing.md`, and added `poetry` to the environment `nix develop` creates.

I'll wait another day or two to see if anyone else has comments. I also think that ultimately @martinvonz should merge this, as he is likely to have to deal with branch protection issues when this is merged. (I don't understand exactly what the best branch protection setup for the `gh-pages` branch is).

martinvonz commented on docs/contributing.md:
```diff
@@ -131,7 +134,62 @@ These are listed roughly in order of decreasing importance.
+To build the website, you must have Python and `poetry` installed. If
+your distribution packages `poetry`, something like `apt install
+python3-poetry` is likely the best way to install it. Otherwise, you
+can download Python form <https://python.org> or follow the [Python
```
s/form/from/

martinvonz commented on docs/contributing.md:
```diff
@@ -131,7 +134,62 @@ These are listed roughly in order of decreasing importance.
+As you edit the `md` files, the website should be rebuilt and reloaded
+in your browser automatically, unless build errors occur.
+
+You should
+occasionally check the terminal from which you ran `mkdocs serve` for
+any build errors or warnings. Warnings about
+`"GET /versions.json HTTP/1.1" code 404` are expected and harmless.
```
nit: reflow text (maybe have your IDE do that)

martinvonz commented on docs/contributing.md:
```diff
@@ -189,6 +194,88 @@ occasionally check the terminal from which you ran `mkdocs serve` for
+2. Make sure `jjfan.github.com/jj` includes the `gh-pages` branch of the jj repo
+and run `git fetch origin gh-pages`.
+
+3. Go to the Github repository settings, enable Github Pages, and configure them
```
nit: Github->GitHub (uppercase H) throughout

ilyagr replied:
I added a commit to make this fix everywhere.

martinvonz reviewed (approved):
Looks great. Thank you!

ilyagr commented:
@martinvonz, feel free to merge this whenever you want.

ilyagr commented:
As expected, the branch protection rules make themselves known.

https://github.com/martinvonz/jj/actions/runs/6003122517/job/16280883112#step:6:77

```
error: failed to push branch gh-pages to origin: "remote: error: GH006: Protected branch update failed for refs/heads/gh-pages.        
remote: error: Changes must be made through a pull request.        
To https://github.com/martinvonz/jj
 ! [remote rejected] gh-pages -> gh-pages (protected branch hook declined)
error: failed to push some refs to 'https://github.com/martinvonz/jj'"
Error: Process completed with exit code 1.
```

martinvonz commented:
I've relaxed the branch protections to only require linear updates now.

ilyagr commented:
I asked it to "rerun failed jobs" -- and I think it worked! 🎉
